### PR TITLE
[move-formatter]Add new module of code formatter

### PR DESCRIPTION
**Pull request by robinlzw**
## Background:
A formatting tool, commonly referred to as a pretty-printer, converts the Abstract Syntax Tree (AST) generated by a parser for a specific language into a well-structured and visually appealing string of code. Currently, there have been discussions within the community regarding the need for a formatting tool specifically tailored for the Move language. However, implementing such a tool poses non-trivial challenges, resulting in the absence of a mature and user-friendly formatting solution for Move.

## Motivation:
To address this gap, MoveBit aims to develop a comprehensive formatting tool for the Move language(movefmt for short). This tool is designed to be versatile, with the capability to operate within command terminal environments. It can also be easily integrated into Aptos-cli or aptos-move-analyzer.

## Description:
This PR adds the doc, code and UT for movefmt.

### CODE REVIEW

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+            type_info::type_of<CoinType>().account_address
+        ).supply
+    }
+
+    spec schema CoinSubAbortsIf<CoinType>{
+        use aptos_framework::optional_aggregator;
+        amount: u64;
+        let addr = type_info::type_of<CoinType>().account_address;
+        let maybe_supply = global<CoinInfo<CoinType>>(addr).supply;
+        include(option::is_some(maybe_supply)) ==> optional_aggregator::SubAbortsIf {
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move:186]: There should be a space after include

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+        let post coin_info = global<CoinInfo<CoinType>>(account_addr);
+        let post supply = option::spec_borrow(coin_info.supply);
+        let post value = optional_aggregator::optional_aggregator_value(supply);
+        let post limit = optional_aggregator::optional_aggregator_limit(supply);
+        modifies global<CoinInfo<CoinType>>(account_addr);
+        aborts_if monitor_supply && parallelizable && !exists<
+            aggregator_factory::AggregatorFactory
+        >(@aptos_framework);
+        /// [managed_coin::high-level-req-2]
+        ensures exists<CoinInfo<CoinType>>(account_addr) && coin_info.name == name && coin_info.
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move:464]: This looks a little bit strange

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move [hunk 3]
```diff
@@ -0,0 +1,1443 @@
...
+    }
+
+    //
+    //  Aggregatable coin functions
+    //
+
+    /// Creates a new aggregatable coin with value overflowing on `limit`. Note that this function can
+    /// only be called by Aptos Framework (0x1) account for now because of `create_aggregator`.
+    public(friend) fun initialize_aggregatable_coin<CoinType>(aptos_framework: &signer)
+        : AggregatableCoin<CoinType>{
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:185]: There should be a space before `{`?

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move [hunk 4]
```diff
@@ -0,0 +1,1443 @@
...
+        );
+    }
+
+    /// This should be called by on-chain governance to update the config and allow
+    /// or disallow upgradability of total supply.
+    public fun allow_supply_upgrades(
+        aptos_framework: &signer,
+        allowed: bool
+    )
+        acquires SupplyConfig {
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:172]: is it a normal case that `acquires` should be in a separate line? cc @vineethk

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 3 above)
Comment (rahxephon89) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:185]: Also, not sure where should we put `:`.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move [hunk 5]
```diff
@@ -0,0 +1,1443 @@
...
+            value == 0,
+            error::invalid_argument(EDESTRUCTION_OF_NONZERO_TOKEN)
+        )
+    }
+
+    /// Extracts `amount` from the passed-in `coin`, where the original token is modified in place.
+    public fun extract<CoinType>(
+        coin: &mut Coin<CoinType>,
+        amount: u64
+    ): Coin<CoinType>{
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:443]: the same space issue with `{`

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move [hunk 6]
```diff
@@ -0,0 +1,1443 @@
...
+    #[test(framework = @aptos_framework, other = @0x123)]
+    #[expected_failure(abort_code = 0x50003, location = aptos_framework::system_addresses)]
+    fun test_supply_initialize_fails(framework: signer, other: signer) {
+        aggregator_factory::initialize_aggregator_factory_for_test(&framework);
+        initialize_with_aggregator(&other);
+    }
+
+    #[test(framework = @aptos_framework)]
+    fun test_supply_initialize(framework: signer)
+        acquires CoinInfo {
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:1250]: Same issue, at the first glance, I thought `acquires CoinInfo` is a statement in the function.

File: third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move [hunk 7]
```diff
@@ -0,0 +1,8 @@
+module test {
+    fun f(x: u64): u64 {
+        if (x == 1) {if (x == 11) { 11 }}
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move:3]: I am not sure whether it is good idea to reformat the `if` block into one line here. cc @vineethk

File: third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    fun make_up_address(x: u64): address {
+        @0x42
+    }
+
+    fun f11()
+        !reads *(0x42), *(0x43) {}
+
+    fun f12()
+        pure {}
+}
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move:55]: I don't think we should remove the last empty line?

File: third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move [hunk 9]
```diff
@@ -0,0 +1,12 @@
+/// test_point: Multiple blank lines between use statements
+
+module BlockLine {
+    // Multiple blank lines between statements
+    use aptos_std::type_info::{
+        /* use_item before */ Self,
+        TypeInfo};
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move:7]: `}` should be in a new line?

File: third_party/move/tools/move-formatter/tests/formatter/tuple/input5.fmt.move [hunk 10]
```diff
@@ -0,0 +1,31 @@
+address 0x42 {
+module example {
+    /**
+    * This function doesn't return anything. It's just for demonstration purposes. */
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/formatter/tuple/input5.fmt.move:4]: This line looks a little bit strange to me.

File: third_party/move/tools/move-formatter/tests/formatter/struct/input4.fmt.move [hunk 11]
```diff
@@ -0,0 +1,21 @@
+module complex_module {
+
+    // Struct with nested comments and complex types
+    struct ComplexStruct2<T, U>{
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/formatter/struct/input4.fmt.move:4]: There should be a space before `{`.

File: third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+    /// The `String` struct holds a vector of bytes that all represent
+    /// valid ASCII characters. Note that these ASCII characters may not all
+    /// be printable. To determine if a `String` contains only "printable"
+    /// characters you should use the `all_characters_printable` predicate
+    /// defined in this module.
+    struct String has /*comment*/ copy, drop /*comment*/, store /*comment*/ {
+        // comment
+        bytes: vector<u8>, // comment
+
```
Comment (rahxephon89) [inline third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move:17]: not sure whether we should remove this empty line? cc @vineethk

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move (hunk 1 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move:186]: Yes, an oversight here.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 3 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:185]: Yes, There should be a space before `{`

File: third_party/move/tools/move-formatter/tests/complex/incentives.fmt.move [hunk 13]
```diff
@@ -0,0 +1,3617 @@
...
+    /// Verify failure for overflow.
+    fun test_range_check_coin_merge() {
+        let target_coins = assets::mint_test<QC>(HI_64); // Mint coins.
+        // Attempt invalid invocation.
+        range_check_coin_merge(1, &target_coins, 12345);
+        assets::burn(target_coins); // Burn target coins.
+    }
+
+    #[test(econia = @econia, integrator = @user)]
+
```
[the remark this answers is not in the document]
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/complex/incentives.fmt.move:2931]: Yes, it makes sense

File: third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move (hunk 8 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move:55]: I have tried rustfmt, and it automatically adds an extra line of whitespace. Can it be like rustfmt, which adds an extra line of whitespace if the last line of the source file is not a blank line, and if there are multiple blank lines at the end of the source file, it retains one line of whitespace?

Comment (rahxephon89) [reply third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move:55]: Yeah, I think we can follow what rustfmt does here.

File: third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move (hunk 9 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move:7]: Yes, it makes sense

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move (hunk 1 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move:186]: Fixed

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 3 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move]: fixed

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 5 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move]: fixed

File: third_party/move/tools/move-formatter/tests/complex/incentives.fmt.move (hunk 13 above)
[the remark this answers is not in the document]
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/complex/incentives.fmt.move]: fixed, you can see third_party/move/tools/move-formatter/tests/complex/input6.fmt.move

File: third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move (hunk 8 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/fun/input5.fmt.move]: fixed

File: third_party/move/tools/move-formatter/tests/formatter/tuple/input5.fmt.move (hunk 10 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/tuple/input5.fmt.move:4]: I can not get your point.

File: third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move (hunk 9 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move:7]: The situation here is caused by comments and blank lines. For more common examples, it is not an issue. You can refer to this new use case: third_party/move/tools/move-formatter/tests/formatter/use/input6.fmt.move.

File: third_party/move/tools/move-formatter/tests/formatter/struct/input4.fmt.move (hunk 11 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/struct/input4.fmt.move]: fixed

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move (hunk 2 above)
Comment (rahxephon89) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.move:464]: What does rust fmt do when breaking long logical expressions?

File: third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move (hunk 7 above)
Comment (rahxephon89) [reply third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move:3]: Does rustfmt do the same thing?

Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move:3]: rustfmt will increase into new lines, regardless of whether the block contains a simple statement or not. We will adjust the processing logic here to be consistent with rustfmt

File: third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move (hunk 9 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/use/input5.fmt.move:7]: Yes, it's difficult, there's no general solution here. But fortunately, in most cases, users would not make such complex use statement blocks. Various other formatting tools may also exhibit unreasonable behavior in such scenarios, which can be observed in their respective issue.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+        let post coin_info = global<CoinInfo<CoinType>>(account_addr);
+        let post supply = option::spec_borrow(coin_info.supply);
+        let post value = optional_aggregator::optional_aggregator_value(supply);
+        let post limit = optional_aggregator::optional_aggregator_limit(supply);
+        modifies global<CoinInfo<CoinType>>(account_addr);
+        aborts_if monitor_supply && parallelizable && !exists<
+            aggregator_factory::AggregatorFactory
+        >(@aptos_framework);
+        /// [managed_coin::high-level-req-2]
+        ensures exists<CoinInfo<CoinType>>(account_addr) && coin_info.name == name && coin_info
```
[the remark this answers is not in the document]
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move:464]: The line-breaking behavior here is consistent with rustfmt. For example, if the line ends with `aa.bb`, it will be split into lines as `aa\n[$indent].bb`

File: third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move (hunk 7 above)
Comment (vineethk) [reply third_party/move/tools/move-formatter/tests/complex/branch/input2.fmt.move:3]: Looks like this is being resolved to what `rustfmt` does, which is likely the better choice.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 4 above)
Comment (vineethk) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:172]: I think this may require a bit more discussion. Given that we are expanding to `reads` and `writes` along with `acquires`, the idea was that we separate all these entities out, otherwise it is hard to parse them (eg., you can say `reads A !writes B reads C, E !writes D`.

So, perhaps if there are more than one kind of resource specifier (`reads`, `writes`, `acquires`), then we need to have them out separated, else not (especially if they fit within column limit)?

I wanted to report this as a bug, but maybe I'll mention it here - we seem to have an issue where you can specify a very long list of resources after `acquires` (like `acquires LongName1, LongName2, ...`) and they all stay in the same line even if they exceed the column limit.

File: third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move (hunk 12 above)
Comment (vineethk) [reply third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move:17]: Yes, there seems to be an issue in formatting structs in general and new lines:

```
struct Blah {

    x: u8,

    y: u64,

    z: u64

}
```

The above will be left unchanged by the formatter.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move (hunk 4 above)
Comment (vineethk) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.fmt.move:172]: In any case, I agree that `acquires` here would be better placed in the previous line.

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+        ensures dst_coin.value == old(dst_coin.value) + source_coin.value;
+    }
+
+    /// An account can only be registered once.
+    /// Updating `Account.guid_creation_num` will not overflow.
+    spec register<CoinType>(account: &signer) {
+        let account_addr = signer::address_of(account);
+        let acc = global<account::Account>(account_addr);
+        aborts_if !exists<CoinStore<CoinType>>(account_addr) && acc.guid_creation_num +
+            2 >= account::MAX_GUID_CREATION_NUM;
```
[the remark this answers is not in the document]
Comment (vineethk) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move:492]: Agree. Looks like taking into account the precedence of operators when doing line breaks would enhance readability here.

[the remark this answers is not in the document]
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move:492]: maybe should like:
```
        aborts_if !exists<CoinStore<CoinType>>(account_addr) &&
            acc.guid_creation_num + 2 >= account::MAX_GUID_CREATION_NUM;
```

File: third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move (hunk 12 above)
Comment (robinlzw) [reply third_party/move/tools/move-formatter/tests/formatter/spec_struct/input3.fmt.move:17]: after formatted, it should like:
```
struct Blah {
    x: u8,
    y: u64,
    z: u64
}
```
?

File: third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move (hunk 15 above)
[the remark this answers is not in the document]
Comment (rahxephon89) [reply third_party/move/tools/move-formatter/tests/aptos_framework_case/coin.spec.fmt.move:492]: Yeah, it looks good.
